### ISSUE TITLE
[FR] improve thread scheduling and affinity

### ISSUE THREAD

**Issue by genivia-inc** [Genivia/ugrep#305]
I was asked by a ugrep user if I had plans to further accelerate the performance of recursive searching by setting thread affinity in ugrep and to improve worker job scheduling, if it helps. So yes, I thought about that and it is relatively easy to do with pthreads (I taught it in [my HPC class at FSU](https://www.cs.fsu.edu/~engelen/courses/HPC/)), _but the pthreads C++11 code is not portable_, so I hesitated to do this until later.

We all know that setting thread affinity can improve threading performance when certain conditions are met in the way threads are used (thread lifetime, RAM/cache use, memory access time versus CPU time ratio).

Running a preliminary test of ugrep with Ubuntu quad core x64 w/ HTT (8 logical cores) shows that a search of /usr with thread affinity runs in 0.12 seconds _which is up to 2x faster than without affinity_. So yes, it is worthwhile to set the thread affinity of the worker thread pool in ugrep. However, MacOS and Windows do not seem to benefit much, if anything.

On Debian and Ubuntu 2.9GHz Intel Core i7 quad core with hyperthreading (8 logical cores) and 16GB 2133MHz LPDDR3  I get around 700% to 750% CPU utilization when searching `ugrep -Ilr zodiaq /usr` in a container in 0.12 seconds (20437 files and 2011 directories, none of them matching `zodiaq`). Now, we all know that CPU percentage is generally meaningless. It could just be running a busy-wait loop for that matter (so it's easy to increase CPU% by doing whatever stuff). The real time (aka. wall clock time) is ultimately the measure of performance to get the work done in parallel. It measures the time elapsed for the work performed with a level of parallelism i.e. when compared to a single CPU performance we get the speed up.

To my surprise, also recursively searching with option `-z` (`--decompress`) runs a lot faster. I was surprised by this, because decompression threads are used in ugrep to feed decompressed streams or plain input (when not compressed) to the search engines of the worker threads, thereby increasing the concurrency of ugrep beyond the number of available physical cores. Still, it looks even better now to make a push to set thread affinity of the worker thread pool in ugrep.

This optimization will be included in the upcoming 4.3.2 release. Obviously, I need more time to test different programmatic ways to set thread affinity and measure the performance impact.

**Comment by genivia-inc** [comment]
Work in progress: benchmark results for Mac PRO M1 10 core with improved thread scheduling looks even more impressive [than before](https://github.com/Genivia/ugrep-benchmarks):

## results for OpenSSL source code repo directory search

grepping `FIXME|TODO` elapsed real time (s)

| search |     -n |    -nr |    -wn |   -wnr |   -win |  -winr |  -wino | -winor |   -wic |  -wicr |   -wil |  -wilr |
| -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: |
|  ugrep |   0.02 |        |   0.02 |        |   0.02 |        |   0.02 |        |   0.03 |        |   0.02 |        |
|     rg |   0.05 |        |   0.06 |        |   0.04 |        |   0.05 |        |   0.04 |        |   0.04 |        |
|     ag |        |   0.04 |        |   0.04 |        |   0.04 |        |   0.04 |        |   0.04 |        |   0.04 |
|  ggrep |        |   0.10 |        |   0.12 |        |   0.17 |        |   0.17 |        |   0.15 |        |   0.15 |

grepping `char|int|long|size_t|void` elapsed real time (s)

| search |     -n |    -nr |    -wn |   -wnr |   -win |  -winr |  -wino | -winor |   -wic |  -wicr |   -wil |  -wilr |
| -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: |
|  ugrep |   0.04 |        |   0.03 |        |   0.04 |        |   0.04 |        |   0.04 |        |   0.03 |        |
|     rg |   0.04 |        |   0.05 |        |   0.05 |        |   0.09 |        |   0.04 |        |   0.04 |        |
|     ag |        |   0.32 |        |   0.23 |        |   0.23 |        |   0.18 |        |   0.05 |        |   0.05 |
|  ggrep |        |   0.16 |        |   0.21 |        |   0.31 |        |   0.43 |        |   0.28 |        |   0.12 |

grepping `ssl-?3(\.[0-9]+)?` elapsed real time (s)

| search |     -n |    -nr |    -wn |   -wnr |   -win |  -winr |  -wino | -winor |   -wic |  -wicr |   -wil |  -wilr |
| -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: |
|  ugrep |   0.02 |        |   0.02 |        |   0.02 |        |   0.02 |        |   0.03 |        |   0.02 |        |
|     rg |   0.06 |        |   0.05 |        |   0.05 |        |   0.04 |        |   0.05 |        |   0.05 |        |
|     ag |        |   0.05 |        |   0.04 |        |   0.04 |        |   0.04 |        |   0.04 |        |   0.04 |
|  ggrep |        |   0.09 |        |   0.10 |        |   0.11 |        |   0.11 |        |   0.09 |        |   0.09 |

## results for Swift source code repo directory search

grepping `_(RUN|LIB|NAM)[A-Z_]+` elapsed real time (s)

| search |     -n |    -nr |    -wn |   -wnr |   -win |  -winr |  -wino | -winor |   -wic |  -wicr |   -wil |  -wilr |
| -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: |
|  ugrep |   0.12 |        |   0.12 |        |   0.12 |        |   0.12 |        |   0.15 |        |   0.11 |        |
|     rg |   0.22 |        |   0.30 |        |   0.22 |        |   0.22 |        |   0.26 |        |   0.26 |        |
|     ag |        |   0.24 |        |   0.21 |        |   0.26 |        |   0.24 |        |   0.23 |        |   0.22 |
|  ggrep |        |   0.42 |        |   0.51 |        |   0.55 |        |   0.55 |        |   0.49 |        |   0.48 |

grepping `String|Int|Double|Array|Dictionary` elapsed real time (s)

| search |     -n |    -nr |    -wn |   -wnr |   -win |  -winr |  -wino | -winor |   -wic |  -wicr |   -wil |  -wilr |
| -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: |
|  ugrep |   0.16 |        |   0.15 |        |   0.18 |        |   0.17 |        |   0.17 |        |   0.13 |        |
|     rg |   0.24 |        |   0.22 |        |   0.26 |        |   0.31 |        |   0.26 |        |   0.20 |        |
|     ag |        |   0.85 |        |   0.49 |        |   0.68 |        |   0.59 |        |   0.24 |        |   0.24 |
|  ggrep |        |   0.54 |        |   0.66 |        |   1.54 |        |   1.77 |        |   1.46 |        |   0.57 |

grepping `(class|struct)\sS[a-z]+T` elapsed real time (s)

| search |     -n |    -nr |    -wn |   -wnr |   -win |  -winr |  -wino | -winor |   -wic |  -wicr |   -wil |  -wilr |
| -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: |
|  ugrep |   0.12 |        |   0.12 |        |   0.12 |        |   0.12 |        |   0.16 |        |   0.11 |        |
|     rg |   0.23 |        |   0.24 |        |   0.23 |        |   0.22 |        |   0.21 |        |   0.22 |        |
|     ag |        |   0.25 |        |   0.22 |        |   0.23 |        |   0.22 |        |   0.22 |        |   0.22 |
|  ggrep |        |   0.55 |        |   0.63 |        |   0.81 |        |   0.81 |        |   0.75 |        |   0.71 |

grepping `for\s[a-z]+\sin` elapsed real time (s)

| search |     -n |    -nr |    -wn |   -wnr |   -win |  -winr |  -wino | -winor |   -wic |  -wicr |   -wil |  -wilr |
| -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: | -----: |
|  ugrep |   0.14 |        |   0.12 |        |   0.12 |        |   0.12 |        |   0.16 |        |   0.11 |        |
|     rg |   0.27 |        |   0.27 |        |   0.23 |        |   0.24 |        |   0.25 |        |   0.25 |        |
|     ag |        |   0.34 |        |   0.26 |        |   0.26 |        |   0.26 |        |   0.22 |        |   0.23 |
|  ggrep |        |   0.51 |        |   0.52 |        |   0.70 |        |   0.70 |        |   0.64 |        |   0.57 |

**Comment by genivia-inc** [comment]
Completed the optimization and tested on MacOS (arm64 and x64), Windows x64, Ubuntu x64, Android termux arm64, RPi3 w/ Debian-based Linux, and Cygwin. This should compile on FreeBSD, _but I have not been able to test with FreeBSD_ since we don't have a machine available to do so.

The thread affinity and priority is set for the calling thread as follows (code subject to change):
~~~
// set this thread's affinity and priority, if supported by the OS, ignore errors to leave scheduling to the OS
static void set_this_thread_affinity_and_priority(size_t cpu)
{
#if defined(OS_WIN) || defined(__CYGWIN__)
  (void)SetThreadAffinityMask(GetCurrentThread(), DWORD_PTR(1) << cpu);
#elif defined(__APPLE__)
  (void)pthread_set_qos_class_self_np(QOS_CLASS_USER_INITIATED, 0);
#elif defined(HAVE_SCHED_SETAFFINITY)
  cpu_set_t cpuset;
  CPU_ZERO(&cpuset);
  CPU_SET(cpu, &cpuset);
  (void)sched_setaffinity(0, sizeof(cpu_set_t), &cpuset);
#elif defined(HAVE_CPUSET_SETAFFINITY)
  cpuset_t cpuset;
  CPU_ZERO(&cpuset);
  CPU_SET(cpu, &cpuset);
  (void)cpuset_setaffinity(CPU_LEVEL_WHICH, CPU_WHICH_PID, getpid(), sizeof(cpuset_t), &cpuset);
#elif defined(HAVE_PTHREAD_SETAFFINITY_NP)
  cpu_set_t cpuset;
  CPU_ZERO(&cpuset);
  CPU_SET(cpu, &cpuset);
  (void)pthread_setaffinity_np(pthread_self(), sizeof(cpu_set_t), &cpuset);
#endif

#if defined(OS_WIN) || defined(__CYGWIN__)
  (void)SetThreadPriority(GetCurrentThread(), THREAD_PRIORITY_HIGHEST);
#elif defined(__APPLE__)
  (void)setpriority(PRIO_DARWIN_THREAD, 0, -20);
#elif defined(HAVE_PTHREAD_SETSCHEDPRIO)
  (void)pthread_setschedprio(pthread_self(), -20);
#elif defined(HAVE_SETPRIORITY)
  (void)setpriority(PRIO_PROCESS, 0, -20);
#endif

  (void)cpu;
}
~~~
The performance impact on MacOS is not observable in my tests, which is not unexpected because MacOS doesn't offer a thread affinity API as far as I know and I've found the thread performance on MacOS quite optimal already. Setting the MacOS thread QoS and priority is probably a good idea anyway, it can't hurt.

**Comment by genivia-inc** [comment]
Included with release v4.3.2.
